File: caffeine_lsp/src/caffeine_lsp_ffi.mjs
import { Ok, Error } from "./gleam.mjs";

export function init_io() {
  return undefined;
}

export function read_line() {
  try {
    const chunks = [];
    const buf = new Uint8Array(1);
    while (true) {
      const n = Deno.stdin.readSync(buf);
      if (n === null || n === 0) return new Error(undefined);
      if (buf[0] === 10) break; // \n
      chunks.push(buf[0]);
    }
    return new Ok(new TextDecoder().decode(new Uint8Array(chunks)));
  } catch (_e) {
    return new Error(undefined);
  }
}

export function read_bytes(n) {
  try {
    const buf = new Uint8Array(n);
    let offset = 0;
    while (offset < n) {
      const bytesRead = Deno.stdin.readSync(buf.subarray(offset));
      if (bytesRead === null || bytesRead === 0) return new Error(undefined);
      offset += bytesRead;
    }
    return new Ok(new TextDecoder().decode(buf));
  } catch (_e) {
    return new Error(undefined);
  }
}

export function write_stdout(data) {
  const buf = new TextEncoder().encode(data);
  let offset = 0;
  while (offset < buf.length) {
    const written = Deno.stdout.writeSync(buf.subarray(offset));
    if (written === null || written === 0) break;
    offset += written;
  }
  return undefined;
}

export function write_stderr(data) {
  const buf = new TextEncoder().encode(data + "\n");
  let offset = 0;
  while (offset < buf.length) {
    const written = Deno.stderr.writeSync(buf.subarray(offset));
    if (written === null || written === 0) break;
    offset += written;
  }
  return undefined;
}

export function rescue(fn) {
  try {
    return new Ok(fn());
  } catch (e) {
    Deno.stderr.writeSync(
      new TextEncoder().encode("LSP handler crashed: " + String(e) + "\n"),
    );
    return new Error(undefined);
  }
}

File: caffeine_lsp/src/caffeine_lsp_ffi.erl
-module(caffeine_lsp_ffi).
-export([init_io/0, read_line/0, read_bytes/1, write_stdout/1, write_stderr/1]).

%% Set stdin/stdout to binary mode for correct byte-level reads.
init_io() ->
    io:setopts(standard_io, [binary]),
    nil.

%% Read a single line from stdin (up to newline).
read_line() ->
    case io:get_line(standard_io, "") of
        eof -> {error, nil};
        {error, _} -> {error, nil};
        Line -> {ok, Line}
    end.

%% Read exactly N bytes from stdin.
read_bytes(N) ->
    case io:get_chars(standard_io, "", N) of
        eof -> {error, nil};
        {error, _} -> {error, nil};
        Data when is_binary(Data) -> {ok, Data};
        Data when is_list(Data) -> {ok, unicode:characters_to_binary(Data)}
    end.

%% Write raw bytes to stdout.
write_stdout(Data) ->
    io:put_chars(standard_io, Data),
    nil.

%% Write to stderr (for logging).
write_stderr(Data) ->
    io:put_chars(standard_error, [Data, "\n"]),
    nil.
